Make prefill in quick open configurable

ColCh opened the pull request:
Close #41725

Demo: https://streamable.com/psso1

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=41586) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=41586) before we can accept your contribution.<br/><br/>:x: ColCh [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=41586)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=41586) it.</sub>

usernamehw commented:
Sorry for my questions, but still...

Have you tried to invoke `quick pick` | `go to symbol` | `command palette` | ... when no editor is opened?
> `Cannot read property 'getControl' of null`

> `Cannot read property 'getControl' of undefined`
--- 

You've already made similar PR in Atom => and you hid functionality under configuration and turned it off by default. What is different now?

---

Multi-line: in your Atom PR when having multi-line selection you don't prefill the input, right? Seems more logical to me to skip multi-line selections here too...

ColCh commented:
> Have you tried to invoke quick pick | go to symbol | command palette | ... when no editor is opened?

wow thank you!  didn't expected this case. Fixed

> You've already made similar PR in Atom => and you hid functionality under configuration and turned it off by default. What is different now?

That behaviour was requested by member of Atom. Personally, I think that this feature should be turned ON by default and be unconfigurable. What do you think?

Webstorm has this feature as well. And AFAIK there is no config to turn it off.

Hm... I think it will be useful to control this feature

I will create 2 config entries:

* to turn it ON/OFF
* to control which prefixes to prefill from selection (it's for symbols and other prefixes I don't know)

> Multi-line: in your Atom PR when having multi-line selection you don't prefill the input, right? Seems more logical to me to skip multi-line selections here too...

Agreed. Skipped that too. Fixed. Thank you!

usernamehw commented on src/vs/workbench/browser/parts/quickopen/quickOpenController.ts:
```diff
@@ -539,6 +541,60 @@ export class QuickOpenController extends Component implements IQuickOpenService
+	}
+
+	private shouldPrefill(prefix?: string): boolean {
+		const isTurnedOn = this.configurationService.getValue('workbench.quickOpen.prefillFromSelection') as string;
```
This is boolean, right?

ColCh replied:
oh yes my bad

fixed

usernamehw commented on src/vs/workbench/browser/parts/quickopen/quickOpenController.ts:
```diff
@@ -539,6 +541,60 @@ export class QuickOpenController extends Component implements IQuickOpenService
+
+		const prefillsPrefix = prefixes.indexOf(prefix) !== -1;
+
+		return prefillsPrefix;
```
Could've been one line? 
`return prefixes.indexOf(prefix) !== -1;`

ColCh replied:
yeah

corrected to be one line

usernamehw commented on src/vs/workbench/electron-browser/main.contribution.ts:
```diff
@@ -226,6 +226,24 @@ configurationRegistry.registerConfiguration({
+				'type': ['string', 'null']
+			},
+			'default': [
+				null,
```
Not obvious for what null stands for.

usernamehw replied:
Then ... what for empty string?

ColCh replied:
it seems that empty string has no reason to be here

removed it

usernamehw replied:
I thought maybe it would be better to remove null from config and use empty string (since empty string is the value in quick picker associated with go to file) - map it to null in code.

bpasero commented:
@ColCh I see two possible solutions:
* we have a setting for this (disabled by default)
* we introduce a new command that opens quick open with the selected text of the editor

My reason why I would never want this as default is that our quick open includes a list of recently opened editors in order of their appearance. This list would not show up anymore as soon as some text is prefilled based on the selection in the editor. 

Btw we have code that gets selection from the editor for the "Open Symbol in Workspace" quick open handler already, check out: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/search/electron-browser/search.contribution.ts#L268

ColCh commented:
@bpasero thanks for suggestions and link to code

Ok, seems reasonable to turn on this feature only for search for all symbols, which is `#` prefix. This also will preserve current behaviour for prefill.

`Open Symbol in Workspace` is cool, but it prefills on `ShowAllSymbolsAction`, not counting all other quick open actions. Also, it selects text in quick open window, which was prefilled.

Feature implemented in current PR prefills on `QuickOpen` level, which stands separately.

I will remove prefill logic from `ShowAllSymbolsAction` and add selection of prefilled value.

ColCh commented:
Done, please check again

ColCh commented:
Build is failing on Travis, but it seems that I can't clearly understand why it's failing...

ColCh commented:
@bpasero ping?

bpasero commented:
This change is not what I was expecting. I think I was expecting either a setting for just quick open (Cmd+P) or just a new command that would do this (I think a new command would be better).

ColCh commented:
Thank you for your thoughts

>  I think I was expecting either a setting for just quick open (Cmd+P) 

This will introduce same logic as symbol search across the project with different implementation. I just think that having single place of handing input prefill is better, isn't it?

> a new command that would do this (I think a new command would be better)

Hm, interesting proposal. You mean, one command will open command menu with defined prefix (quick open, symbol search, other ...) and will append selected text, and another command, which will just open command menu without prefill (like now) ?

Ok, I will code this. AFAIK only symbol search across the project now prefills selected text?

Any other points to consider?

bpasero commented:
@ColCh yeah we could add a new quick open command similar to Cmd+P today that puts the selection in. If there is any code to share between open symbols and open file, that would be good.

bpasero commented:
@ColCh you could also do this from an extension, see my comment in https://github.com/Microsoft/vscode/issues/10413#issuecomment-363735174

ColCh commented:
@bpasero ooh thank you so much for that snippet! It seems that this functionality should not exist in core, but in extension, right?

bpasero commented:
@ColCh yeah let's start with an extension 👍
